DSDEGP-1579: refactor common code

pshapiro4broad opened the pull request:
### Description

DSDEGP-1579: clio code refactoring

This is an intermediate PR to cover changes made so far, there will be more changes made before this is merged to master.

- replace classes GvcfUtil, WgsUbamUtil with ClassUtil
- add transfer model objects for Index and Key
- change code to use HttpCredentials instead of OAuth2BearerToken
- rename addXXX() client APIs to upsertXXX()

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: Ensure that you have added or updated tests
- [ ] **Submitter**: If you're adding/switching libraries or otherwise changing the design, update the [design documentation](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=114531509).
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
* Review cycle:
  * Team may comment on PR at will
  * **Reviewer assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to reviewer** for further feedback
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Move the JIRA issue to QA once this checklist is completed

pshapiro4broad committed:
DSDEGP-1579: replace classes GvcfUtil, WgsUbamUtil with ClassUtil

pshapiro4broad committed:
DSDEGP-1579: refactor ClioWebClient to remove some duplicate code

- add transfer model objects for Index and Key
- change code to use HttpCredentials instead of OAuth2BearerToken
- rename addXXX() client APIs to upsertXXX()

pshapiro4broad committed:
DSDEGP-1579: eliminate custom getSchema APIs

pshapiro4broad committed:
DSDEGP-1579: remove all gvcf/ubam specific APIs from ClioWebClient

danxmoran commented on clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/TransferIndex.scala:
```diff
@@ -0,0 +1,29 @@
+  def jsonSchema: Json
+}
+
+final case class GvcfIndex() extends TransferIndex {
```
Since these don't take parameters, they can be `case object`s

pshapiro4broad replied:
I knew there was something I was missing here. I was also thinking of using enums, but wasn't sure where this class was going yet.

pshapiro4broad committed:
DSDEGP-1579: use a case object for index objects

pshapiro4broad committed:
DSDEGP-1579: refactor Add for gvcf, ubam, into common code

jacarey commented on clio-client/src/main/scala/org/broadinstitute/clio/client/webclient/ClioWebClient.scala:
```diff
@@ -59,101 +56,51 @@ class ClioWebClient(
     dispatchRequest(
-      HttpRequest(uri = "/api/v1/wgsubam/schema")
-        .addHeader(Authorization(credentials = bearerToken))
+      HttpRequest(uri = s"/api/v1/${transferIndex.urlSegment}/schema")
```
Not in scope for this review but at some point we will want to pass in version.

pshapiro4broad replied:
Yeah I'm not 100% sure how versioning will work, we may need different types for input/query/output if we change the schema in an incompatible way (and want to maintain support for v1 APIs)

jacarey commented on clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/TransferWgsUbamV1Key.scala:
```diff
@@ -6,3 +6,8 @@ case class TransferWgsUbamV1Key(flowcellBarcode: String,
+    extends TransferKey {
+
+  def getUrlPath: String =
+    s"$flowcellBarcode/$lane/$libraryName/$location"
```
I almost feel like the location should be the first path segment similar to gvcfs. That way we are consistent with increasing specificity as you move down the segments. location -> flowcell -> lane -> library

pshapiro4broad replied:
That would be OK with me, doing it now would be better than later. What do you think, @tbl3rd and @danxmoran ?

danxmoran replied:
+1

tbl3rd replied:
Generally: I think we should review this PR for maintainability regressions (and easy improvements) only and approve it otherwise.

Specifically: I see $location as different in kind from $barcode, $lane, and $library. I like moving it first because it is going to be shared across all of our indexes. And another issue just occurred to me: Some indexes will have multiple paths, but only one $location, so ...

tbl3rd reviewed (approved):
So much red.

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/commands/ClioCommand.scala:
```diff
@@ -89,6 +107,12 @@ object ClioCommand extends ClioParsers {
+  val deletePrefix = "delete-"
+
   // Names for WGS uBAM commands.
   val getWgsUbamSchemaName = "get-schema-wgs-ubam"
```
This could / should be `s"${getSchemaPrefix}-${WgsUbamIndex.commandName}"`. Similarly for the others.

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/commands/ClioCommand.scala:
```diff
@@ -21,6 +21,16 @@ import shapeless.CNil
+
+sealed trait AddCommand extends ClioCommand {
+  def index: TransferIndex
+  def key: TransferKey
```
If we really want to lock this down, we could add another `type` param to `TransferIndex` for key type, and then this could be `def key: index.KeyType`.

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/AddExecutor.scala:
```diff
@@ -0,0 +1,50 @@
+      .left
+      .map { err =>
+        new RuntimeException(
+          s"Invalid metadata given at $location. Run the '${ClioCommand.getSchemaPrefix}${index.commandName}' command to see the expected JSON format for ${index.name}s.",
```
Should there be a `-` between `${ClioCommand.getSchemaPrefix}` and `${index.commandName}`?

pshapiro4broad replied:
No, the prefixes end in `-`

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/util/ClassUtil.scala:
```diff
@@ -0,0 +1,12 @@
+package org.broadinstitute.clio.client.util
+
+object ClassUtil {
```
Extreme nitpick: this is generic enough that it might belong in the clio-util project.

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/webclient/ClioWebClient.scala:
```diff
@@ -59,101 +56,51 @@ class ClioWebClient(
-    input: TransferWgsUbamV1Key,
-    transferWgsUbamV1Metadata: TransferWgsUbamV1Metadata
-  )(implicit bearerToken: OAuth2BearerToken): Future[HttpResponse] = {
+  def upsert[T](transferIndex: TransferIndex, key: TransferKey, metadata: T)(
```
Instead of `T`, can we use `transferIndex.metadataType`?

danxmoran replied:
Actually, I think this won't work in current Scala unless we break the params into two lists because of how scoping for arguments works:
```scala
def upsert(transferIndex: TransferIndex)(key: transferIndex.KeyType, metadata: transferIndex.MetadataType)
```
Up to you if you want to try locking it down.

danxmoran commented on clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/TransferIndex.scala:
```diff
@@ -0,0 +1,65 @@
+
+  def commandName: String
+
+  type metadataType
```
Nitpick: even for abstract type members, I think the convention is to use upper camel case.

pshapiro4broad replied:
OK

danxmoran commented on clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/TransferIndex.scala:
```diff
@@ -0,0 +1,65 @@
+
+case object GvcfIndex extends TransferIndex {
+
+  override def urlSegment: String = "gvcf"
```
Nitpick: I think all these `override def`s should work as `override val`s

pshapiro4broad replied:
I'll give it a try

danxmoran reviewed (approved):
Some parting suggestions, but overall looks great! Much better than it was.

pshapiro4broad committed:
DSDEGP-1579: more refactorings

- finish refactoring Query
- switch ordering of location in wgs web service to be the first path element, instead of the last
- do partial work to refactor Delete, not complete

pshapiro4broad committed:
DSDEGP-1579: code review feedback

- type names start with uppercase
- move ClassUtil into clio-util project
-

pshapiro4broad committed:
Merge remote-tracking branch 'origin/develop' into ps_DSDEGP-1579_refactor_common_code

# Conflicts:
#	clio-server/src/test/scala/org/broadinstitute/clio/server/service/PersistenceServiceSpec.scala

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/commands/ClioCommand.scala:
```diff
@@ -31,19 +47,24 @@ case object GetServerVersion extends ClioCommand
 
-@CommandName(ClioCommand.getWgsUbamSchemaName)
-case object GetSchemaWgsUbam extends ClioCommand
+@CommandName(ClioCommand.getSchemaPrefix + WgsUbamIndex.commandName)
```
Why not still `ClioCommand.getWgsUbamSchemaName`? Similarly for the rest.

pshapiro4broad replied:
I'll change this back, I was trying out getting rid of the `ClioCommand` members for each command but in the end it didn't seem worthwhile (yet). One can imagine a future version of the code where commands and index names are passed separately as command line arguments, and the command line parser is initialized programmatically.

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/commands/ClioCommand.scala:
```diff
@@ -21,6 +21,22 @@ import shapeless.CNil
+
+sealed trait QueryCommand extends ClioCommand {
+  def index: TransferIndex
+  def queryInput: Any
```
Could you add a `FIXME` or `TODO` comment here?

pshapiro4broad replied:
Yes

pshapiro4broad committed:
Merge remote-tracking branch 'origin/develop' into ps_DSDEGP-1579_refactor_common_code

pshapiro4broad committed:
DSDEGP-1579: last (?) round of feedback
